flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/nn/QuadTree.scala

```diff
@@ -0,0 +1,340 @@
+ * efficient KNN queries and generalizing to n-dim.
+ *
+ * @param minVec vector of the corner of the bounding box with smallest coordinates
+ * @param maxVec vector of the corner of the bounding box with smallest coordinates
+ * @param distMetric metric, must be Euclidean or squareEuclidean
+ * @param maxPerBox threshold for number of points in each box before slitting a box
+ */
+class QuadTree(minVec: Vector, maxVec: Vector, distMetric: DistanceMetric, maxPerBox: Int){
+
+  class Node(center: Vector, width: Vector, var children: Seq[Node]) {
+
+    val nodeElements = new ListBuffer[Vector]
+
+    /** for testing purposes only; used in QuadTreeSuite.scala
+      *
+      * @return center and width of the box
+      */
+    def getCenterWidth(): (Vector, Vector) = {
+      (center, width)
+    }
+
+    def contains(queryPoint: Vector): Boolean = {
+      overlap(queryPoint, 0.0)
+    }
+
+    /** Tests if queryPoint is within a radius of the node
+      *
+      * @param queryPoint
+      * @param radius
+      * @return
+      */
+    def overlap(queryPoint: Vector, radius: Double): Boolean = {
+      var count = 0
+      for (i <- 0 to queryPoint.size - 1) {
+        if (queryPoint(i) - radius < center(i) + width(i) / 2 &&
+          queryPoint(i) + radius > center(i) - width(i) / 2) {
+          count += 1
+        }
+      }
+
+      if (count == queryPoint.size) {
+        true
+      } else {
+        false
+      }
+    }
+
+    /** Tests if queryPoint is near a node
+      *
+      * @param queryPoint
+      * @param radius
+      * @return
+      */
+    def isNear(queryPoint: Vector, radius: Double): Boolean = {
+      if (minDist(queryPoint) < radius) {
+        true
+      } else {
+        false
+      }
+    }
+
+    /**
+     * used in error handling when computing minDist to make sure
+     * distMetric is Euclidean or SquaredEuclidean
+     * @param message
+     */
+    case class metricException(message: String) extends Exception(message)
+
+    /**
+     * minDist is defined so that every point in the box
+     * has distance to queryPoint greater than minDist
+     * (minDist adopted from "Nearest Neighbors Queries" by N. Roussopoulos et al.)
+     *
+     * @param queryPoint
+     * @return
+     */
+
+    def minDist(queryPoint: Vector): Double = {
+      var minDist = 0.0
+      for (i <- 0 to queryPoint.size - 1) {
+        if (queryPoint(i) < center(i) - width(i) / 2) {
+          minDist += math.pow(queryPoint(i) - center(i) + width(i) / 2, 2)
+        } else if (queryPoint(i) > center(i) + width(i) / 2) {
+          minDist += math.pow(queryPoint(i) - center(i) - width(i) / 2, 2)
+        }
+      }
+
+      if (distMetric.isInstanceOf[SquaredEuclideanDistanceMetric]) {
+        minDist
+      } else if (distMetric.isInstanceOf[EuclideanDistanceMetric]) {
+        math.sqrt(minDist)
+      } else{
+        throw metricException(s" Error: metric must be Euclidean or SquaredEuclidean!")
```

Comment:
Why we need a new exception type? I think `IllegalArgumentException` could be used in this case.